### ISSUE TITLE
Allow "favorite" videos (heart icon)

### ISSUE THREAD

**Issue by whyboris** [whyboris/Video-Hub-App#763]
Seems like an easy thing to add and some people might prefer this over the 5-star rating system.

_Proposal:_
A "favorite" video, indicated by a heart icon in the top-left of a thumbnail (not shown on a _folder_ when in _folder view_), is equivalent to a 5-star rating. This means any 5 star video automatically has the heart on it, and removing the heart from it removes the 5-star rating. No additional metadata field will be added to the video.

_Importantly_ there will be a toggle to hide the hearts -- letting those who don't want to see them to not see the hearts.

Should be an easy add 👍

**Comment by Sooyoung98** [comment]
Hello! My team and I would like to work on this issue for our [class assignment](https://cmu-313.github.io//assignments/hw6). Would it be possible? Thank you!

**Comment by whyboris** [comment]
Happy to help out with this any time 🤝 

When I wrote "Should be an easy add 👍" it was a note for myself 😅 and it will be trickier for someone not familiar with the codebase, even though I tried to organize it well :trollface: 

Don't hesitate to reach out with questions - I can help with general suggestions or specific ones (as fine-grained as you request) - just let me know 😎

**Comment by whyboris** [comment]
@Sooyoung98 - let me know if you run into the same problem as #779 (the solution is in the thread there, but I'd like to know if it happens to you too) 🤝

**Comment by Sooyoung98** [comment]
In fact we did run into the problem, but happened to solve it by using **npm install --force** ! 
We will definitely take reference to it if this error pops up again though! 

We really appreciate your support :)

**Comment by sunnyliang6** [comment]
@whyboris Hi Boris! I'm working on toggling the heart icon and have questions. If the video is already favorite'd / 5 stars and the heart icon is clicked, we should remove the heart but what should we then change the star rating to? Perhaps 0? Also, how might I get and set metadata (rating) from the ts files in the `views` directory?

**Comment by whyboris** [comment]
Hello @sunnyliang6 - I think that when a heart is added, regardless of the number of stars the video had we should make the number of stars 5. When the heart is clicked again, we should make the star rating go to "not rated" which happens to be 0.5

```

**Comment by whyboris** [comment]
Getting the metadata should be possible by simply viewing the `ImageElement` object's `stars` property 🤔 

Meanwhile, setting it from any component may be possible with:

```ts
    this.imageElementService.HandleEmission({
      index: this.video.index,
      stars: rating
    });
```
See: https://github.com/whyboris/Video-Hub-App/blob/main/src/app/components/meta/meta.component.ts#L127

Please feel free to reach out any time!

**Comment by Sooyoung98** [comment]
Hey boris! @whyboris  some of my teammates and I are trying to look into the backend side of the things for this issue, and just wanted to double check on what your initial expectations were on any changes on the backend side?

Also, @sunnyliang6  is one of our team members, just letting you know in case of any confusion :)

**Comment by whyboris** [comment]
Thank you 😊 

The favorite (heart) property for a video would be filled in temporarily when a hub is opened up, and deleted before the `.vha2` file is saved. Both happen on the "back end" (in _Node_) 🤝 

The insertion happens here:
https://github.com/whyboris/Video-Hub-App/blob/main/node/main-support.ts#L544

The deleting happens here:
https://github.com/whyboris/Video-Hub-App/blob/main/node/main-support.ts#L224

Cheers!

**Comment by Sooyoung98** [comment]
@whyboris I see! 

Based on what we got so far from code-tracing and what you commented above, we are currently thinking of tackling on the stars attribute.

In the insertion part, we are thinking of placing the stars value as 5 
In the deletion part, we are thinking of setting the stars value back to 0.5 

**I just wanted to double-check on this with you to see if what we are currently thinking is going on the right track, as you are visioning.** 
Would this align with your intended result?

Again, thank you so much for your support! Appreciate it :)

**Comment by whyboris** [comment]
When opening a hub or closing the hub you do not need to touch the `stars` at all.

You should add a key to the _Optional_ fields here:
https://github.com/whyboris/Video-Hub-App/blob/main/interfaces/final-object.interface.ts#L45
(you can name it `favorite`, `heart`, or something else you think is fitting).
This field will be the one you add before the hub is open and delete before it is closed 👍 

ps - stars value of 5 is not allowed, please update to 5.5 for a 5-star-rated video:
https://github.com/whyboris/Video-Hub-App/blob/main/interfaces/final-object.interface.ts#L1

**Comment by Sooyoung98** [comment]
Oh I see! Got it I understood the process :) 

Thank you so much for such a detailed explanation. 
My team and I will work on this more and let you know if there's any more questions! 

Thank you! Your support means a lot 😄

**Comment by Sooyoung98** [comment]
@whyboris Hey Boris! We have a question for you regarding this issue. 
After committing the changes I made which solves "The favorite (heart) property for a video would be filled in temporarily when a hub is opened up, and deleted before the .vha2 file is saved.",  @sunnyliang6 and I realized that even without the changes I made @sunnyliang6 's commits already  have an implementation of the favorites feature. 

This is my commit : https://github.com/Sooyoung98/313-Add-Favorites/commit/e6046134f26992c38922b077e6cb00eb1cbbbbf9

This is Sunny's commit : https://github.com/Sooyoung98/313-Add-Favorites/commit/5847c19e829d4a7c04b406c8c2e7c4b91a671ad0

Given the fact that the feature works fine even without the favorites attribute on the metadata, do you still want me to keep my changes just in case you might need that in the future?

**Comment by whyboris** [comment]
I see how the back-end changes may not have been needed 🤔 
The field being present could have made it easier to have a "see only favorites" toggle, but under the hood it could just show all the videos with a 5-star rating (no need for the extra field added in / removed) 🤔 

I'm happy either way you choose to go - feel free to keep the back-end change and make use of it (no need to `ngOnInit` fill it in then), or discard the back-end changes. 

It looks like you got the feature to work great - thank you for the contribution!

Feel free to finish with whatever you have and open a PR as is. I'm comfortable doing clean-up (if any is needed). 

If you feel like throwing in more pieces of the feature, consider adding a `settingsButton` that will toggle whether the heart icon shows up or is hidden 👍 - if so, please add some text in English in `en.json` and I'll take care of all the other language translations 🤝 

https://github.com/whyboris/Video-Hub-App/blob/main/src/app/common/settings-buttons.ts#L251

**Comment by sunnyliang6** [comment]
@whyboris Hi Boris, an update on @Sooyoung98's question. We've realized that we need the `favorite` attribute. We updated the `handleEmission` with this new attribute so that, on the frontend, when the stars are changed and when the heart is changed, both the `stars` and the `favorite` values are updated correctly in the backend. 👍

Now, our issue is that the UI is not being updated accordingly. When the stars icon are clicked, the heart icon is not being updated. When the heart icon is clicked, the stars icons are not being updated. This is because the `starRatingHack` variable is not being updated when the heart icon is clicked and our new `heartLitHack` variable is not being updated when the stars icons are clicked.

We've identified that the frontend code responsible for the stars is located in `meta.component.html`. In `details.component.html`, the stars are rendered in the view through the `app-meta-item` element. However, the heart icon must be nested in all classes called `video-box` in order for it to be rendered on the video thumbnails correctly.

We've implemented a work-around fix (https://github.com/Sooyoung98/313-Add-Favorites/commit/b5bccada8cd264579161ce201e44d226cce2a544) for heart to star, meaning that when we toggle on the heart, the star UI can change accordingly. We did this by creating a new input field in `app-meta-item` called `starRatingHack`. (The key point here is that `starRatingHack` and `heartLitHack` are the variables responsible for actually changing the UI.) However, we cannot find a working solution for star to heart. 😔

(Note: `heartLitHack` is newly defined here: https://github.com/Sooyoung98/313-Add-Favorites/commit/5847c19e829d4a7c04b406c8c2e7c4b91a671ad0)

How would you suggest we change `heartLitHack` in order to get the star to heart UI update to work?

One work-around we've thought of is to move the location of the heart off the thumbnail to be able to put the heart in the metadata instead of the `video-box` and avoid this issue.

I apologize for the long, detailed message. Thank you so much for your continued support! 🙂

**Comment by whyboris** [comment]
The UI doesn't update sometimes when it should, so I opt for a hack: `this.cd.detectChanges()` (it's all over my files 😅 ) -- see if that fixes the problem 🤞 

I'll look tomorrow some more at your specific commits and see if I have any other ideas. I'm off to bed 🛏️ 😴

**Comment by joshHuang02** [comment]
Hi @whyboris, I am also working with @Sooyoung98 and @sunnyliang6 on this issue.
Just to update so far:
- we have added the 'show favorites' setting into the view settings page and the top ribbon, should work just like the other view settings
- added heart icon for all the views, which will show if 'show favorite' setting is toggled
- below changes were only added to details view so far, plan to add to all views once we figure things out
- added code to set heart appearance depending on rating
- clicking heart icon changes its appearance, sets the rating (0.5 or 5.5) in meta.component, and updates the star rating visual (using starRatingHack) upon click 

Problem: 
- heart icon appearance correctly sets depending on rating, but we can't figure out how to trigger the update upon clicking a star
- `this.heartLitHack = this.video.stars == 5.5;` → this line in [details.component.ts](https://github.com/whyboris/Video-Hub-App/blob/97e476f3a6d64b9a93fed60dd4de5f3fd077b8db/src/app/components/views/details/details.component.ts) will correctly set heart appearance whenever it is run, like in `ngOnInit()`
- unsuccessful in trying to get the click star in meta.components to trigger stuff in details.component
- tldr heart icon updates rating correctly but changing rating does not trigger heart icon update :(

[up to date changes here](https://github.com/whyboris/Video-Hub-App/commit/97e476f3a6d64b9a93fed60dd4de5f3fd077b8db)

We have identified the problem as the heart icon is located in each of the views components while the rating events are in meta.component. I can get this to work if there is a way to make clicking a star trigger something in the view component ts files to update the heart icon. Do you have any advice?

**Comment by joshHuang02** [comment]
update again: I found how to do it with ngOnChanges()

**Comment by whyboris** [comment]
Closed with #796 👍 

Thank you again for your contribution @Sooyoung98, @sunnyliang6, @anitacheng19, and @joshHuang02
